File: hierarchal.cpp
#include<iostream>
using namespace std;

class Marvellous 
{
    public:
        int C;
        Marvellous()
        {
            C = 30;
            cout<<"Inside Marvellous Constructor\n";
        }
        ~Marvellous()
        {
            cout<<"Inside Marvellous Destructor\n";
        }
        void Sun()
        { cout<<"Inside Sun of MArvellous\n"; }

};

class Demo : public Marvellous
{
    public:
        int A;
        Demo()
        {
            A = 10;
            cout<<"Inside Demo Constructor\n";
        }
        ~Demo()
        {
            cout<<"Inside Demo Destructor\n";
        }
        void Fun()
        { cout<<"Inside Fun of Demo\n";}
        

};

class Hello : public Marvellous
{
    public:
        int B;
        Hello()
        {
            B = 20;
            cout<<"Inside Hello Constructor\n";
        }
        ~Hello()
        {
            cout<<"Inside Hello Destructor\n";
        }
        void Gun()
        { cout<<"Inside Gun of Hello\n";}

};


int main()
{
    Demo dobj;
    Hello hobj;
    


    return 0;
}

File: Hierarchal.java

class Base 
{
    public int A,B;
    public Base()
    {
        System.out.println("Base Constructor\n");
    }
    public void Fun()
    {
        System.out.println("Inside Base Fun\n");
    }

}


class Derived extends Base
{
    public int X,Y;
    public Derived()
    {
        System.out.println("Derived Constructor\n");
    }
    public void Gun()
    {
        System.out.println("Inside Derived Gun\n");
    }

}

class DerivedX extends Base
{
    public int P,Q;
    public DerivedX()
    {
        System.out.println("DerivedX Constructor\n");
    }
    public void Sun()
    {
        System.out.println("Inside DerivedX Sun");
    }
}


class Hierarchal
{
    public static void main(String args[])
    {
       Base dobj = new DerivedX();
       dobj.Fun();
       

    }
}
